crates/verifier/src/converter.rs

```diff
@@ -0,0 +1,135 @@
+use core::cmp::Ordering;
+
+use bn::{AffineG1, AffineG2, Fq, Fq2};
+
+use crate::{
+    constants::{CompressedPointFlag, MASK},
+    error::Error,
+};
+
+/// Checks if the first byte is zero and the rest of the bytes are zero.
+pub(crate) fn is_zeroed(first_byte: u8, buf: &[u8]) -> bool {
+    if first_byte != 0 {
+        return false;
+    }
+    for &b in buf {
+        if b != 0 {
+            return false;
+        }
+    }
+
+    true
+}
+
+/// Deserializes an Fq element from a buffer.
+///
+/// If this Fq element is part of a compressed point, the flag that indicates the sign of the
+/// y coordinate is also returned.
+pub(crate) fn deserialize_with_flags(buf: &[u8]) -> Result<(Fq, CompressedPointFlag), Error> {
+    if buf.len() != 32 {
+        return Err(Error::InvalidXLength);
+    };
+
+    let m_data = buf[0] & MASK;
+    if m_data == u8::from(CompressedPointFlag::Infinity) {
+        if !is_zeroed(buf[0] & !MASK, &buf[1..32]) {
+            return Err(Error::InvalidPoint);
+        }
+        Ok((Fq::zero(), CompressedPointFlag::Infinity))
+    } else {
+        let mut x_bytes: [u8; 32] = [0u8; 32];
+        x_bytes.copy_from_slice(buf);
+        x_bytes[0] &= !MASK;
+
+        let x = Fq::from_be_bytes_mod_order(&x_bytes).expect("Failed to convert x bytes to Fq");
+
+        Ok((x, m_data.into()))
+    }
+}
```

Comment:
The interface for `is_zeroed` can be simplified to do the mask inside of the loop, and the logic for the loop can also be simplified with:

```rust
    first_byte == 0 && buf.iter().all(|&b| b == 0)
```